lambda/translmod.ml

```diff
@@ -919,40 +928,136 @@ let required_globals ~flambda body =
   Translprim.clear_used_primitives ();
   required
 
+let add_arg_block_to_module_block primary_block_lam size restr =
+  let primary_block_id = Ident.create_local "*primary-block*" in
+  let arg_block_id = Ident.create_local "*arg-block*" in
+  let arg_block_lam =
+    apply_coercion Loc_unknown Strict restr (Lvar primary_block_id)
+  in
+  let get_field i = Lprim (mod_field i, [Lvar primary_block_id], Loc_unknown) in
+  let all_fields = List.init size get_field @ [Lvar arg_block_id] in
+  let arg_block_field = size in
+  let new_size = size + 1 in
+  Llet(Strict, layout_module, primary_block_id, primary_block_lam,
+       Llet(Strict, layout_module, arg_block_id, arg_block_lam,
+            Lprim(Pmakeblock(0, Immutable, None, alloc_heap),
+                  all_fields,
+                  Loc_unknown))),
+  new_size,
+  Some arg_block_field
+
+let add_parameters lam params =
+  let params =
+    List.map
+      (fun name ->
+        { name;
+          layout = Pvalue Pgenval;
+          attributes = Lambda.default_param_attribute;
+          mode = Lambda.alloc_heap })
+      params
+  in
+  let inline =
+    (* We want to inline the functor so that [-instantiate] compiles away the
+       function call and actually substitutes. *)
+    Always_inline
+  in
+  lfunction ~kind:(Curried { nlocal = 0 }) ~params ~return:(Pvalue Pgenval)
+    ~attr:{ default_function_attribute with is_a_functor = true; inline }
+    ~loc:Loc_unknown
+    ~body:lam
+    ~mode:alloc_heap
+    ~ret_mode:alloc_heap
+    ~region:true
+
+let transl_implementation_module ~scopes module_id (str, cc, cc2) =
+  let path = global_path module_id in
+  let lam, size =
+    transl_struct ~scopes Loc_unknown [] cc path str
+  in
+  match cc2 with
+  | None -> lam, size, None
+  | Some cc2 -> add_arg_block_to_module_block lam size cc2
+
+let wrap_toplevel_functor_in_struct code =
+  Lprim(Pmakeblock(0, Immutable, None, Lambda.alloc_heap),
+        [ code ],
+        Loc_unknown)
+
+(* Convert an flambda-style implementation (module block only) to a
+   non-flambda-style one (set global as side effect) *)
+
+let wrap_in_setglobal implementation =
+  let code =
+    Lprim (Psetglobal implementation.compilation_unit, [implementation.code],
+           Loc_unknown)
+  in
+  { implementation with code }
+
 (* Compile an implementation *)
 
 type compilation_unit_style =
   | Plain_block
   | Set_global_to_block
   | Set_individual_fields
 
-let transl_implementation_plain_block compilation_unit (str, cc) =
+let has_parameters () =
+  !Clflags.parameters <> []
```

Comment:
I'm surprised that we need this. Can this information be passed to functions instead?

Reply:
Yes, or at any rate we can ask `Env` for it. Fixing. (I think at some point the code must have been less uniform in handling parameters so sometimes `Env.parameters` didn't have the right information. The tests are passing still so that time must be in the past.)